### PR TITLE
chore(amplify_api): add support for apiName to GraphQL requests

### LINKED ISSUE #547: API (GraphQL) - Support multiple auth modes
**Describe the bug**
amplify-flutter is not supporting multiple auth modes because the **amplify_api package** is not receiving neither propagating a parameter called "apiName".

**apiName** is used by the amplify-ios/amplify-flutter to infer which api to use.

If we configure multiple auth modes in **amplifyconfiguration.dart**, all the api operations failed returning this error:

**More than one graphQL API configured. Could not infer which API to call, recoverySuggestion: Use the apiName to specify which API to call** 

**To Reproduce**
1. Configure multiple auth modes. (https://docs.amplify.aws/lib/graphqlapi/authz/q/platform/flutter)
2. Send an operation (query, mutation, subscription)

**Expected behavior**
We expect that GraphQLRequest receive the apiName in order to be able to configure the multi auth mode.

**Platform**
Amplify Flutter current supports iOS and Android.  This issue is reproducible in (check all that apply):
[x] Android
[x] iOS

### PR DESCRIPTION
**Pull request by jodafm**
*Issue #, if available:*
Resolves #547 

*Description of changes:*
1. add apiName parameter in GraphQLRequest.dart 
2. propagate the apiName to queries, mutations and subscriptions

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by odemkovychromexsoft** [comment]
I wonder if somebody can provide some planning for this pull request. I need to use different GraphQL endpoints and can't do that without this changes. I  just hope it will be merged and released soon.

**Comment by jodafm** [comment]
I found that is necessary to configure a awsCredentialsProvider in the APIPlugin, specifically in the Android side. If not all the requests through an API with **IAM auth type** fail with the following error:
Screen_Shot_2021-05-17_at_11 12 16_AM

I am going to update the PR with the fix.

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/AmplifyApiPlugin.kt
Comment (ragingsquirrel3) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/AmplifyApiPlugin.kt:74]: request: Could the changes in this file please be removed from the PR? I see the comment about adding to support IAM mode, but strictly speaking, I do not believe these changes are needed for the purpose of the PR: multiple auth modes. I pulled down the PR, omitted this file, and it seemed to work with an app which I configured to have multiple auth modes with API keys and cognito user pools.

Side note: The change in this file, or something similar might be needed for IAM mode but it needs some more research.

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiRequest.kt
Comment (ragingsquirrel3) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterApiRequest.kt:118]: question: Would it make sense to omit the "SHARED" line here? I think this will only be used for GraphQL, correct? REST also uses an `apiName` but I think it is nested in another object and made available above by calling `getApiPath` (not sure why it has that name).

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt
Comment (ragingsquirrel3) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt:61]: Nice job reusing this callback! #DRY

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt
Comment (ragingsquirrel3) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt:157]: nit: Is the `!!` necessary given there is a null check already?

**Review by ragingsquirrel3 — CHANGES_REQUESTED** [review]
1) Thanks for pointing out this issue.
2) 100x thanks for submitting this PR.
3) Thanks for your patience in awaiting review.

Other than the android config changes, would it be possible to include some unit tests for this change? If that's going to be difficult or time-consuming, I might be able to provide some help by pseudo-coding them in comments or something.

File: packages/amplify_api_plugin_interface/lib/src/GraphQL/GraphQLRequest.dart
Comment (jodafm) [reply packages/amplify_api_plugin_interface/lib/src/GraphQL/GraphQLRequest.dart:45]: I think the null check is necessary because the apiName is an optional value, so if the value is null we want to avoid set it in the serialized map. If not, I experienced an unexpected behavior in the native platforms.

**Comment by jodafm** [comment]
@ragingsquirrel3 thank you for the review.
I'm going to be working on your requests in the next hours.

**Comment by jodafm** [comment]
@ragingsquirrel3 let me know if you need additional changes.
On the other hand, do you recommend me to create another issue for the IAM auth mode support?

**Comment by ragingsquirrel3** [comment]
Thanks for adding tests, will review them as soon as possible.

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt
Comment (ragingsquirrel3) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt:61]: nit: is there an extra space here before `=`? If so, can it removed?

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt
Comment (ragingsquirrel3) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt:33]: Nit: can this extra newline be removed?

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt
Comment (ragingsquirrel3) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt:72]: same nit here for extra space

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt
Comment (ragingsquirrel3) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt:134]: same nit for space

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt
Comment (ragingsquirrel3) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt:145]: same nit for space

File: packages/amplify_api_plugin_interface/lib/src/GraphQL/GraphQLRequest.dart
Comment (ragingsquirrel3) [reply packages/amplify_api_plugin_interface/lib/src/GraphQL/GraphQLRequest.dart:45]: Yes, I believe you are correct as I was able to cause an ios error when I removed this check. thanks.

**Review by ragingsquirrel3 — CHANGES_REQUESTED** [review]
This is looking very close! I left another comment about tests. I am also going to find someone else on our side to provide a 2nd review as this PR should have 2 approvals before merge.

Request: Could the title of the PR (and eventual commit message) please be changed to "chore(amplify_api): add support for apiName to GraphQL requests" ? I think it will be more precise and informative. There are some other things to check before we can 100% say we support multi auth modes, like updating the documentation and testing how the CLI works with this. That is not needed in this PR, just explaining why this message might make more sense.

**Comment by ragingsquirrel3** [comment]
@jodafm yes, if you don't mind, please create another issue for the android IAM config issue, although it looks like you already raised an issue there and they have a fix https://github.com/aws-amplify/amplify-android/pull/1351 . An issue in this (flutter) repo can communicate the status of this to others until that fix is shipped on android side, integrated here, and tested. Thanks for communicating to that team!

**Comment by jodafm** [comment]
@ragingsquirrel3.
I made the changes requested.
Let me know if we need fix something else.

**Comment by ragingsquirrel3** [comment]
Thanks for updating the tests. I will review them as soon as possible, and I'm also trying to get a 2nd reviewer, as I mentioned.

**Comment by jodafm** [comment]
Thanks for your help @ragingsquirrel3

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt
Comment (ragingsquirrel3) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt:212]: nit: still an extra space here

File: packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt
Comment (ragingsquirrel3) [inline packages/amplify_api/android/src/main/kotlin/com/amazonaws/amplify/amplify_api/FlutterGraphQLApi.kt:221]: nit: space here, too

**Review by ragingsquirrel3 — APPROVED** [review]
Thanks, approving with small nits on spacing, assigned @fjnoyp as second reviewer whenever he gets a chance.

**Comment by jodafm** [comment]
> Thanks, approving with small nits on spacing, assigned @fjnoyp as second reviewer whenever he gets a chance.

Thank you, I removed the spaces that you mentioned .

**Comment by ragingsquirrel3** [comment]
@jadafm thanks again for patience. I assure you this is still on our radar.

File: packages/amplify_api_plugin_interface/lib/src/GraphQL/GraphQLRequest.dart
Comment (fjnoyp) [inline packages/amplify_api_plugin_interface/lib/src/GraphQL/GraphQLRequest.dart:34]: Is this null check necessary?  We made the check above because variables = {}; but here apiName is null to begin with.

**Review by fjnoyp — APPROVED** [review]
Looks great, thanks for the PR!  Just one minor comment and otherwise looks good to me.

**Comment by matthewfx** [comment]
Hi! Any ETA on that issue? It would be highly appreciated if you could merge it. :)

**Review by matthewfx — APPROVED** [review]

**Comment by ragingsquirrel3** [comment]
@jodafm, merged! Thanks again for your patience and contribution to this important addition to the library. I added a comment on https://github.com/aws-amplify/amplify-flutter/issues/547 about this.

**Comment by jodafm** [comment]
@ragingsquirrel3 thank you!

**Pull request merged**
